File: python/dataminr_pair_coding.py
"""
Parse a string that has phone ids and their quantities.
IDs can be either for Android or IPhone.
Android format is A + 4 characters.
IPhone format is I + 3 characters. 
"""
import re
from itertools import takewhile, dropwhile, izip

EXAMPLE = "Aa1133Iab2Aa112Iac3"

def tokens(s):
    pattern = r'A\w{3}|I\w{2}|\d+' 
    return (g.group() for g in re.finditer(pattern, s))

def tokens_to_pairs(tokens):
    i = iter(tokens)
    return ((ident, int(quant)) for (ident, quant) in izip(i, i))

def parse_into_map(s):
    d = {}

    for (k, v) in tokens_to_pairs(tokens(s)):
        d.update({k: d.get(k, 0) + v})

    return d

print parse_into_map(EXAMPLE)

File: scala/src/main/scala/DataMinrPairCoding.scala

/**
  * Parse a string that has phone ids and their quantities.
  * IDs can be either for Android or IPhone.
  * Android format is A + 4 characters.
  * IPhone format is I + 3 characters.
  */
object DataMinrPairCoding extends App {
  val exampleString = "Aa1133Iab2Aa112Iac3"

  type Tokens = String

  def tokens(s: String): Iterator[Tokens] = raw"A\w{3}|I\w{2}|\d+".r.findAllIn(s)

  def parseIntoMap(s: String) = {
    tokens(s).grouped(2).map {
      case List(a, b) => (a, b.toInt)
    }
    .toStream
    .groupBy(_._1)
    .mapValues(_.map(_._2).sum)
  }

  println(s"${parseIntoMap(exampleString)}")
}
